internal/controllers/s3userclaim/provisioner.go

```diff
@@ -311,3 +342,73 @@ func (r *Reconciler) assembleCephUserSecret(userName, secretName string) (*corev
 
 	return secret, nil
 }
+
+func (r *Reconciler) generateSubUserAccessMap(desiredSubUsers []string,
+	currentSubUsers []admin.SubuserSpec) map[string]string {
+	// Create a hashmap to move all spec and ceph subUsers to it
+	subUserFullIdAccessMap := make(map[string]string)
+
+	// Tag specSubUsers with "create"
+	for _, subUser := range desiredSubUsers {
+		cephSubUserFullId := generateSubUserFullId(r.cephUserFullId, subUser)
+		subUserFullIdAccessMap[cephSubUserFullId] = consts.SubUserTagCreate
+	}
+
+	// Add read-only subUser to subUsers to prevent if from removing
+	subUserFullIdAccessMap[r.readonlyCephUserFullId] = consts.SubUserTagCreate
+
+	// Tag cephSubUsers as remove if they are not already in the hashmap and remove them otherwise
+	// since they are already available on ceph and not needed to created.
+	for _, currentSubUser := range r.cephUser.Subusers {
+		_, exists := subUserFullIdAccessMap[currentSubUser.Name]
+		if exists {
+			delete(subUserFullIdAccessMap, currentSubUser.Name)
+		} else {
+			subUserFullIdAccessMap[currentSubUser.Name] = consts.SubUserTagRemove
+		}
+	}
+	return subUserFullIdAccessMap
+}
+
+func (r *Reconciler) generateSubUser(ctx context.Context, cephUserFullId string,
+	desiredSubuser admin.SubuserSpec) error {
+	r.logger.Info(fmt.Sprintf("Create subUser: %s", desiredSubuser.Name))
+	if err := r.rgwClient.CreateSubuser(ctx, admin.User{ID: cephUserFullId}, desiredSubuser); err != nil {
+		r.logger.Error(err, "failed to create subUser")
+		return err
+	}
+	return nil
+}
+
+func (r *Reconciler) RemoveSubuserAndSecret(ctx context.Context, cephUserFullId string,
+	subuserToRemove admin.SubuserSpec) error {
+	err := r.rgwClient.RemoveSubuser(ctx, admin.User{ID: cephUserFullId}, subuserToRemove)
+	r.logger.Info(fmt.Sprintf("Remove subUser: %s", subuserToRemove.Name))
+	if err != nil {
+		r.logger.Error(err, "failed to remove subUser")
+		return err
+	}
+	// Extrace subUser name from the subUserFullId
+	subUser, err := extractSubUserName(subuserToRemove.Name)
+	if err != nil {
+		r.logger.Error(err, "failed to remove s3SubUserSecret")
```

Comment:
I think the error message could be improved to better describe what has happened.

Reply:
There is a log in the `extractSubUserName` function: `cannot parse the cephSubUserFullId`
So I thought that if anything goes wrong, we should see the mentioned log and then being unable to delete the subuser secret.